src/librustc/middle/const_eval.rs

```diff
@@ -1273,97 +1032,152 @@ fn resolve_trait_associated_const<'a, 'tcx: 'a>(tcx: &'a ty::ctxt<'tcx>,
+        Bool(b) => cast_const_int(tcx, Infer(b as u64), ty),
+        Float(f) => cast_const_float(tcx, f, ty),
+        Char(c) => cast_const_int(tcx, Infer(c as u64), ty),
+        _ => Err(CannotCast),
     }
 }
 
-fn lit_to_const(lit: &ast::Lit, ty_hint: Option<Ty>) -> ConstVal {
-    match lit.node {
-        ast::LitStr(ref s, _) => Str((*s).clone()),
-        ast::LitByteStr(ref data) => {
-            ByteStr(data.clone())
-        }
-        ast::LitByte(n) => Uint(n as u64),
-        ast::LitChar(n) => Uint(n as u64),
-        ast::LitInt(n, ast::SignedIntLit(_, ast::Plus)) => Int(n as i64),
-        ast::LitInt(n, ast::UnsuffixedIntLit(ast::Plus)) => {
-            match ty_hint.map(|ty| &ty.sty) {
-                Some(&ty::TyUint(_)) => Uint(n),
-                _ => Int(n as i64)
+fn lit_to_const<'tcx>(lit: &ast::Lit_,
+                      tcx: &ty::ctxt<'tcx>,
+                      ty_hint: Option<Ty<'tcx>>,
+                      ) -> Result<ConstVal, ConstMathErr> {
+    use syntax::ast::*;
+    const I8MAX: u64 = ::std::i8::MAX as u64;
+    const I16MAX: u64 = ::std::i16::MAX as u64;
+    const I32MAX: u64 = ::std::i32::MAX as u64;
+    const I64MAX: u64 = ::std::i64::MAX as u64;
+    const U8MAX: u64 = ::std::u8::MAX as u64;
+    const U16MAX: u64 = ::std::u16::MAX as u64;
+    const U32MAX: u64 = ::std::u32::MAX as u64;
+    const U64MAX: u64 = ::std::u64::MAX as u64;
+    match *lit {
+        LitStr(ref s, _) => Ok(Str((*s).clone())),
+        LitByteStr(ref data) => Ok(ByteStr(data.clone())),
+        LitByte(n) => Ok(Integral(U8(n))),
+        LitInt(_, SignedIntLit(_, Minus)) => unreachable!(),
+        LitInt(n @ 0...I8MAX, SignedIntLit(TyI8, _)) => Ok(Integral(I8(n as i8))),
+        LitInt(n @ 0...I16MAX, SignedIntLit(TyI16, _)) => Ok(Integral(I16(n as i16))),
+        LitInt(n @ 0...I32MAX, SignedIntLit(TyI32, _)) => Ok(Integral(I32(n as i32))),
+        LitInt(n @ 0...I64MAX, SignedIntLit(TyI64, _)) => Ok(Integral(I64(n as i64))),
+        LitInt(n, SignedIntLit(TyIs, _)) => {
+            Ok(Integral(Isize(try!(ConstIsize::new(n as i64, tcx.sess.target.int_type)))))
+        },
+
+        LitInt(_, SignedIntLit(ty, _)) => Err(ConstMathErr::LitOutOfRange(ty)),
+
+        LitInt(n, UnsuffixedIntLit(sign)) => {
+            match (ty_hint.map(|t| &t.sty), sign) {
+                (Some(&ty::TyUint(_)), Minus) => panic!("unary neg on signed int"),
+                (Some(&ty::TyInt(ity)), sign) => {
+                    lit_to_const(&LitInt(n, SignedIntLit(ity, sign)), tcx, ty_hint)
+                },
+                (Some(&ty::TyUint(uty)), _) => {
+                    lit_to_const(&LitInt(n, UnsignedIntLit(uty)), tcx, ty_hint)
+                },
+                (None, Plus) => Ok(Integral(Infer(n))),
+                (None, Minus) => Ok(Integral(InferSigned(-(n as i64)))),
+                (Some(&ty::TyEnum(ref adt, _)), _) => {
+                    let hints = tcx.lookup_repr_hints(adt.did);
```

Comment:
addressed `FIXME` for enum repr type

https://github.com/oli-obk/rust/commit/928be878ee9ffb303d3a82106d582bc82ec65acc#commitcomment-15587252